File: src/features/typing/data/ts/palindrome_partitioning.ts
function partition(s: string): string[][] {
    const result: string[][] = [];
    const backtrack = (start: number, current: string[]): void => {
        if (start === s.length) {
            result.push([...current]);
            return;
        }
        for (let end = start + 1; end <= s.length; end++) {
            const piece = s.slice(start, end);
            if (isPalindrome(piece)) {
                current.push(piece);
                backtrack(end, current);
                current.pop();
            }
        }
    };
    backtrack(0, []);
    return result;
}

function isPalindrome(s: string): boolean {
    for (let i = 0, j = s.length - 1; i < j; i++, j--) {
        if (s[i] !== s[j]) {
            return false;
        }
    }
    return true;
}

File: src/features/typing/data/dart/palindrome_partitioning.dart
class Solution {
  List<List<String>> partition(String s) {
    final result = <List<String>>[];
    void backtrack(int start, List<String> current) {
      if (start == s.length) {
        result.add(List.of(current));
        return;
      }
      for (int end = start + 1; end <= s.length; end++) {
        final piece = s.substring(start, end);
        if (_isPalindrome(piece)) {
          current.add(piece);
          backtrack(end, current);
          current.removeLast();
        }
      }
    }
    backtrack(0, <String>[]);
    return result;
  }

  bool _isPalindrome(String s) {
    for (int i = 0, j = s.length - 1; i < j; i++, j--) {
      if (s.codeUnitAt(i) != s.codeUnitAt(j)) {
        return false;
      }
    }
    return true;
  }
}
